regex/src/main/scala/regex/Regex.scala

```diff
@@ -6,30 +6,78 @@ package regex
   * data structure definitions for regular languages
   */
 
-// Add your definitions here
+trait RegularLanguage()
+
+case object Empty   extends RegularLanguage
+case object Epsilon extends RegularLanguage
+case class Character(val c: Char) extends RegularLanguage
+case class Union(val lang1: RegularLanguage, val lang2: RegularLanguage)
+                                                extends RegularLanguage
+case class Concat(val lang1: RegularLanguage, val lang2: RegularLanguage)
+                                                extends RegularLanguage
+case class Star(val lang: RegularLanguage)  extends RegularLanguage
+
+
 
 /** *****************************************************************************
   * Derivatives
   *
   * Fill in the function definitions below
   */
 
-/** Simplifies a regular language */
-def simplify(lang: RegularLanguage): RegularLanguage = ???
+/** Simplifies a regular language. Implemented Recursively. */
+def simplify(lang: RegularLanguage): RegularLanguage =
+  lang match
+    case Union(lang1, lang2) =>
+      if      lang1 == Empty then lang2
+      else if lang2 == Empty then lang1 
+      else Union(simplify(lang1), simplify(lang2))
+
+    case Concat(lang1, lang2) => 
+      if lang1 == Empty || lang2 == Empty then Empty
+      else if lang1 == Epsilon then lang2
+      else if lang2 == Epsilon then lang1
+      else Concat(simplify(lang1), simplify(lang2))
+
+    case Star(subLang) =>
+      if      subLang == Empty   then  Empty
+      else if subLang == Epsilon then Epsilon
+      else Star(simplify(subLang))
+
```

Comment:
I took a different approach to solving this one, but I think that I might like your solution better. I like your use of if and else statements, it makes your code organized/easy to read.